File: implementation/scipamato/scipamato-common-persistence-jooq/src/main/java/ch/difty/scipamato/common/persistence/JooqSortMapper.java
package ch.difty.scipamato.common.persistence;

import java.util.Collection;

import org.jooq.Record;
import org.jooq.SortField;
import org.jooq.impl.TableImpl;

import ch.difty.scipamato.common.entity.ScipamatoEntity;
import ch.difty.scipamato.common.persistence.paging.Sort;

/**
 * Implementations of this interface map sorting specifications into the jOOQ
 * specific {@link SortField}s.
 *
 * @author u.joss
 *
 * @param <R>
 *            the type of the record, extending {@link Record}
 * @param <T>
 *            the type of the entity, extending {@link ScipamatoEntity}
 * @param <TI>
 *            the type of the table implementation of record {@code R}
 */
@FunctionalInterface
public interface JooqSortMapper<R extends Record, T extends ScipamatoEntity, TI extends TableImpl<R>> {

    /**
     * Maps spring data {@link Sort} specifications of table {@code TI} into the
     * jOOQ specific {@link SortField}s.
     *
     * @param sortSpecification
     *            the spring data {@link Sort}
     * @param table
     *            {@code TI}
     * @return collection of {@link SortField}s
     */
    Collection<SortField<T>> map(Sort sortSpecification, TI table);

}

File: common/common-persistence-jooq/src/main/kotlin/ch/difty/scipamato/common/persistence/JooqSortMapper.kt
package ch.difty.scipamato.common.persistence

import ch.difty.scipamato.common.entity.DbEntity
import ch.difty.scipamato.common.entity.ScipamatoEntity
import ch.difty.scipamato.common.persistence.paging.Sort
import org.jooq.Record
import org.jooq.SortField
import org.jooq.Table

/**
 * Implementations of this interface map sorting specifications into the jOOQ specific [SortField]s.
 *
 * @param [R] the type of the record, extending [Record]
 * @param [T] the type of the entity, extending [ScipamatoEntity]
 * @param [T] the type of the table implementation of record [R]
 */
@Deprecated("replace with JooqDbSortMapper")
fun interface JooqSortMapper<R : Record, T : ScipamatoEntity, TI : Table<R>> {

    /**
     * Maps spring data [sortSpecification] of [table] [TI] into the jOOQ specific [SortField]s.
     */
    fun map(sortSpecification: Sort?, table: TI): Collection<SortField<T>>
}

/**
 * Implementations of this interface map sorting specifications into the jOOQ specific [SortField]s.
 *
 * @param [R] the type of the record, extending [Record]
 * @param [T] the type of the entity, extending [DbEntity]
 * @param [T] the type of the table implementation of record [R]
 */
fun interface JooqDbSortMapper<R : Record, T : DbEntity, TI : Table<R>> {

    /**
     * Maps spring data [sortSpecification] of [table] [TI] into the jOOQ specific [SortField]s.
     */
    fun map(sortSpecification: Sort?, table: TI): Collection<SortField<T>>
}
